Ports of release 8.0 to the Mac, Solaris-x86, and Visual Studio 2015.

MichaelJCaruso opened the pull request:
A pull request that:

- brings release 8.0 to MacOS Sierra.
- brings release 8.0 to the Intel version of Solaris 11.3.
- insures that this version of 8.0 builds successfully on Windows 10 using Visual Studio 2015 (the community edition).

Significant changes include:

- updates to the Solaris code base to use Solaris' platform independent atomic memory operations.
- updates to the Linux and Solaris code base to prefer the use of the DCE compliant libuuid UUID generator where available.
- changes to our Visual Studio project files that allow Vision to be built directly from the directory structure found in this repository instead of first needing to rearrange that structure (project files now go to the code instead of requiring that the code first go the project files).
- many, but far from all, of the changes needed to update this code to play well with newer (and in some cases, buggier) versions of the C++ compiler and libraries.

Not included in this pull request is the integration of these changes into release 8.1; however, a git-based integration that does that has been successfully prototyped.  That prototype also gives us a simple transition to git-branch based development and release work-flows but that's a subject for future pull requests :-)

Commit:
Begin the port of Vision to Solaris 11.3 x86.

MichaelJCaruso committed:
Merge branch 'staging' into solaris-11.3-x86

MichaelJCaruso committed:
Option to disable batchvision's automatic memory allocation limit.

To be a good citizen in Linux and Solaris production environments,
batchvision has long imposed a 2gb private memory allocation limit
on itself unless it detects that an explicit limit has already been
set (unix/linux getrlimit RLIMIT_DATA != RLIM_INFINITY).  With this
change, batchvision's self-limiting behavior can also be disabled by
setting the 'VisionDataLimitDisabled' environment variable (to any
value).

MichaelJCaruso committed:
Option to disable 2gb throttle of unlimited data ulimit.

To make batchvision behave properly in production environments,
'batchvision' automatically reduces the amount of private DATA
memory it allows itself to consume to 2gb when started with no
restriction (unix/linux setrlimit RLIMIT_DATA == RLIM_INFINITY).
Historically, that restriction is removed if an explicit limit
other than unlimited (infinity) is set before 'batchvision' is
started.  With this change, the restriction can also be removed
by setting the 'VisionDataLimitDisabled' environment variable.

MichaelJCaruso committed:
Merge branch '8.1-segv-at-exit' into solaris-11.3-x86

MichaelJCaruso committed:
Merge branch '8.1-segv-at-exit' into staging

MichaelJCaruso committed:
Merge branch 'setrlimit-changes' into staging

MichaelJCaruso committed:
Duplicate ad5c6309e763f9112e3c138beeb4b3fb101b6f4c changes under 8.1.

MichaelJCaruso committed:
Merge branch 'master' into staging

MichaelJCaruso committed:
Merge branch 'master' into solaris-11.3-x86

MichaelJCaruso committed:
Merge branch '8.1-bootstrap-fix' into staging

MichaelJCaruso committed:
Merge branch 'staging' into solaris-11.3-x86

MichaelJCaruso committed:
Merge branch 'master' into staging

MichaelJCaruso committed:
Merge branch 'staging' into solaris-11.3-x86

Commit:
Add missing #endif

MichaelJCaruso committed:
VkUUID interface modernization.

Modernize the VkUUID interface and its definition of our portable uuid_t
type to take advantage of the availability of libuuid on modern linux
and unix (e.g., Solaris) systems.  Specifically, our portable uuid_t type
has been moved from the global namespace to namespace V and no longer
attempts to make any pretense of direct coerce-ability to any operating
system type of the same name and intent.  The interface changes made and
propagated here lay the groundwork for the necessary changes to VkUUID's
implementation in the commit(s) to follow.

MichaelJCaruso committed:
Use libuuid to generate uuids on linux systems.

MichaelJCaruso committed:
Rollback VkUUID.{cpp,h} changes on solaris-11.3-x86 branch.

MichaelJCaruso committed:
Merge VkUUID libuuid support into solaris modernization.

MichaelJCaruso committed:
Merge remote-tracking branch 'refs/remotes/origin/solaris-11.3-x86' into solaris-11.3-x86

Commit:
Add -m64 to Solaris x86 compiler options.

MichaelJCaruso committed:
Remove Solaris symbol equivalences not needed in x86 builds.

A bug in some version of the Solaris c++ compiler occasionally caused the
compiler to generate different name manglings for same object, resulting
in undefined symbol linker errors for some references to those objects.
The workaround for that problem is to explicitly equivalence the mangled
names using #pragma weak directives (see the code for a full explanation).
Some of those equivalences are no longer needed for x86 builds using
Solaris Studio 2.1 (they remain for the benefit of sparc builds which
still appear to need them).

Commit:
A working version of vision on x86 Solaris 11.3.

Made possible by the use of Solaris atomic_ops and libuuid, both of
which are also likely to exist in the the SPARC versions of Solaris.

MichaelJCaruso committed:
Create initial visionBuilder MacOS configuration files.

MichaelJCaruso committed:
Merge remote-tracking branch 'origin/solaris-11.3-x86' into macport

MichaelJCaruso committed:
Add Mac (aka Darwin) compiler to list of supported compilers.

MichaelJCaruso committed:
Use libuuid to implement uuid's on Mac.

MichaelJCaruso committed:
Update guess for an initial Mac (Darwin) Vp.h.

MichaelJCaruso committed:
Disable atomic memory operations in Mac port for now.

Temporarily use the no-atomics implementation of atomic memory operations
in the Mac port to allow the port to focus on the big picture of the Mac
runtime and compilation environment.  A future commit will return to this
to make use of the Mac specific version of atomic memory operations (see
kern/OSAtomic.h).

MichaelJCaruso committed:
Apple uses posix mutexes.

MichaelJCaruso committed:
Remove 'll' from '%llp' format specifier.

MichaelJCaruso committed:
VString 'int'->'size_t' and 'clang' related cleanups.

While addressing the 'clang' complaint described below, several
variables that were defined to be 'int's were when they should have been
'size_t's were found and corrected.

'clang' warns when the test in an 'if' or 'while' statement uses an
assignment (i.e., 'var = expr') to simultaneous capture and test the
value of an expression.  This commit addresses that complaint.

MichaelJCaruso committed:
Fix 'clang' if/while assignment warnings.

'clang' warns when the test in an 'if' or 'while' statement uses an
assignment (i.e., 'var = expr') to simultaneous capture and test the
value of an expression.  This commit addresses that complaint.

MichaelJCaruso committed:
Use VkRegExp on Mac.

MichaelJCaruso committed:
Mac signal handling for the Vk.h based code.

MichaelJCaruso committed:
Modern versions of <math.h> prefer 'isfinite' over BSD 'finite'.

MichaelJCaruso committed:
Incremental changes to M_Darwin/make.cc* files.

MichaelJCaruso committed:
Use 'cp -RP' instead of 'cp -d' to copy symlinks.

It seems that 'cp -R -P' may be the more portable way to tell 'cp' to copy
symbolic links (as opposed to the files they reference) and that 'cp -d'
may only work on Linux (at least as far as testing on Solaris 11.3 and the
Mac is concerned).

MichaelJCaruso committed:
Merge branch 'solaris-11.3-x86' into macport

MichaelJCaruso committed:
Merge branch 'master' into solaris-11.3-x86

MichaelJCaruso committed:
Merge branch 'solaris-11.3-x86' into macport

MichaelJCaruso committed:
'isfinite' lives in a different header on Solaris.

'isfinite' seems to be preferred everywhere but Solaris.  It exists there as well,
but including it requires the use of a different header and language standard (c99).
All things considered, it seems easier to continue to use 'finite' from <math.h> on
Solaris.

Commit:
Perfunctory changes to unsupported code for the benefit of 'clang'.

MichaelJCaruso committed:
'long's are 64-bits wide in 64-bit builds.

Don't use 'long long' - it confuses the compiler :-)

Commit:
Shared libraries are named differently on the Mac.

Shared (dynamic) libraries are named and work slightly differently
on the Mac.  To begin with, their names end in '.dylib', not '.so'
or '.so.#', so the first step in getting them to build on the Mac
is changing their 'visionBuilder' target names to remove those
platform dependent suffixes.  To keep 'git' happy, this commit
handles the file renaming part of a two part change.  The next
commit will actually make the required content changes.

Commit:
Shared libraries are named differently on the Mac.

Remove '.so.1' suffixes from our library target names.

Commit:
Shared libraries are named differently on the Mac.

With the machine dependent suffixes removed from the names of our
platform independent shared library names, this commit updates our
platform specific builder templates to add the appropriate suffixes
back to the actual targets being built.

Commit:
Adjust compiler options used to build Mac executables.

Commit:
Shared libraries also WORK differently on the Mac.

On the Mac, shared libraries need to have an installation path
baked into them when they are linked.  That path can be relative
to the executable or library that's loading them but it must
include the name of the library, not just the path to the
directory that contains it.  For example, the command that builds
a 'libV' that can be found by other executables and libraries in
our directory tree needs to include the option:

	-install_name @loader_path/../lib/libV.dylib

Finding a way to generate that option in a platform-independent
makefile turns out to be a little harder than it seems, so for
now, this commit hacks that capability into 'VLINK-CC'.

Commit:
Remove an ill-formed Vca_VcaError from the Vca build.

VcaError uses a long deprecated model for implementing interfaces.  It is not
used anywhere in this code base and, because it fails to provide some required
virtual overrides, it doesn't compile on the Mac and probably wouldn't work even
it did.  Remove it.

Commit:
'extern' template instantiations cause missing typeinfo symbols on the Mac.

Commit:
Enable appropriate existing conditional code on the Mac.

MichaelJCaruso committed:
Linux/Solaris shared libraries need '.so' suffix, not '.so.1'.

Commit:
.gitignore extra files generated by 'c++ -g' on the Mac.

Commit:
Remove limit on debug info produced by '-g' on the Mac.

Commit:
Fix an error in V_VAtomicMemoryOperations_Disabled builds.

Commit:
Use @executable_path relative install_name so our debug builds work.

Customizing VLINK-CC to set -install_name for our dynamic libraries on the
Mac really does need to move into the makefile (and probably use @rpath
relative paths) before the Mac port is done.

Commit:
Remove -v (verbose) flag from Mac library builds.

MichaelJCaruso committed:
Set Mac -install_name using a make recipe, not VLINK-CC

This commit undoes the hack of using VLINK-CC to set a hardwired '-install_name'
in our Mac libraries, introducing a new 'linkerrule' meta-tag to do the job
properly using 'make' (See any of the 'make.*.lib' files in this commit for an
example).

This commit also removes the 'make.rule.*' make recipe files for all of our
shared libraries.  With the introduction of 'linkerrule', the handling of target
directory creation in VLINK-CC where it belongs, and the changing of our
shared library suffix to 'so' instead of 'so.1' on Linux and Solaris, the
custom (and identical) recipes contained in those files are no long needed.

MichaelJCaruso committed:
Merge 8.1 '-U_FORTIFY_SOURCE' 'master' changes into 'macport'

MichaelJCaruso committed:
Merge 8.1 '-U_FORTIFY_SOURCE' 'master' changes into 'solaris-11.3-x86'

MichaelJCaruso committed:
Null merge of branch 'solaris-11.3-x86' into macport

MichaelJCaruso committed:
Stub in a new section for the Mac.

MichaelJCaruso committed:
Merge remote-tracking branch 'origin/solaris-11.3-x86' into macport

MichaelJCaruso committed:
Working Mac implementation of our atomic memory operations.

This commit implements our atomic memory operations on the Mac using what
turns out to be the newly deprecated MAC <libkern/OSatomic.h> atomic API.
This is interesting because the declarations in this header have been
deprecated in favor of C++11's std::atomic (<atomic>) API.  While this is
more work than it's worth as part of this initial Mac port, it is worth it
longer term.  At the moment, <libkern/OSatomic.h> (man atomic_deprecated)
with its function call interface is a better match for our current trait-based
manipulation of un-encapsulated atomic POD values.  In contrast, the proposed
C++11 alternative uses classes to encapsulate atomic POD values, making it
more object oriented but less of a drop-in fit for our current implementation.
Still:

    - It's part of a platform and architecture independent standard that's
      already been implemented for Linux, the Mac, probably Windows, and,
      maybe even someday, Solaris.
    - We already encapsulate atomic values in classes that are a natural fit
      (VCounter, VReferenceable, VSpinlock, etc.) for the std::atomic API,
      making the effort largely one of careful update, not major rewrite.
    - we can get rid of the '-Wno-deprecated-declarations' compiler option
      from the Mac build :-)

MichaelJCaruso committed:
Add omitted 'defined(__APPLE__)' to VkMemory::Destroy.

Additional auditing of the Mac port is probably in order to ensure that there
aren't any more missing or mis-directed conditional compilations.  Still, with
this change, this Mac version can successfully and correctly bootstrap a new
database, making it officially useful and usable !!!

MichaelJCaruso committed:
A simple source file mapper

MichaelJCaruso committed:
Update source paths in *.vcxproj files

MichaelJCaruso committed:
Why not stash these in git - they can always be deleted

MichaelJCaruso committed:
Merge branch 'visual-studio-project-file-rewrite' into macport-plus

MichaelJCaruso committed:
A few fixes on the way to getting 8.0 compiling on Windows.

Including:
    - a long standing error in an 'interlockedExchange' VAtomicMemoryOperation we've never used
    - a couple of missing UCast's in VkUUID.cpp

MichaelJCaruso committed:
Add missing external->windows-internal uuid_t casts to VkUUID.cpp.

MichaelJCaruso committed:
Merge VkUUID::uuid_t windows type casts from 'solaris-11.3-x86' into macport.

MichaelJCaruso committed:
Merge branch 'macport' into macport-plus

MichaelJCaruso committed:
Move Windows platform specific headers (Vp*_Win32_*.h) to src/M_Windows directory.

MichaelJCaruso committed:
Find source files in 8.0 Visual Studio projects using a property page.

No need to move source files and, in a change that should be duplicated in our 'nix/nux builds,
move and rename the Windows Vp*_Win32_1x.h platform specific headers to a .../8.0/src/M_Windows
directory where they too can be found using the Paths.prop property page as well.

MichaelJCaruso committed:
Visual Studio 2015 requires spaces in literal string concatenations.

Hope this doesn't mess up the other compilers :-)

MichaelJCaruso committed:
Move VkLongUnsigneds into V on Windows.

MichaelJCaruso committed:
Fix typo in V::VAtomicMemoryOperations_.h affecting Windows build.

MichaelJCaruso committed:
Add 'vquery.h' to Visual Studio headers project.

MichaelJCaruso committed:
Use <stdint.h> for integer type (eg, int8_t) on Windows.

MichaelJCaruso committed:
Remove working files created during Visual Studio path re-writing.

VCommitter commented on software/src/8.0/src/kernel/VkUUID.cpp:
```diff
@@ -58,16 +31,43 @@ typedef caddr_t ifc_buf_t;
+
+#elif defined(__APPLE__) || defined(__linux__) || defined(sun)
+#define USING_LIBUUID_DCE
+#include <uuid/uuid.h>
```
This isn't necessarily builtin on Linux:

```
../kernel/VkUUID.cpp:58:23: fatal error: uuid/uuid.h: No such file or directory
 #include <uuid/uuid.h>
```

is it your intention to make libuuid-devel a requirement?

MichaelJCaruso replied:
That ought to be a reasonable requirement.  libuuid is part of the Linux runtime and should provide higher quality UUID generation than what we've got.  I'm not sure why RedHat left out the part (libuuid-devel) that contains the include files.  They seem to be present elsewhere.

Don't know how much I'll be able to look, but what object files/libraries are being implicated by the linker?  Is this a clean build?

VCommitter reviewed (commented):
ran across the issue with libuuid-devel.  

Ran into an another problem on Solaris 10:

```
cg error (as) : "/tmp/ccfe.22374.1.s", line 24928 : redefinition of symbol "__1cBVbDVReferenceableImplementation_4n0ALThreadModelFMulti__DRTT_"
gmake: *** [Vca_VStockDeviceFactory.o] Error 1
```

`-DUSING_LEGACY_SPARC_ATOMICS` does fix this error for me.

Didn't try windows yet.

VCommitter commented:
I lied.  `-DUSING_LEGACY_SPARC_ATOMICS` does not fix the redefinition error

MichaelJCaruso committed:
Fix field width/buffer size inconsistencies exposed by Mac runtime buffer checks.

Mac OS runtime buffer checking discovered a very small number of inconsistencies between
actual buffer sizes and the 'sscanf' format string field widths used to write to those
buffers.  The specific issue was the requirement that buffers used to receive 'scanf'
parsed strings have size at least 1 element larger than the 'scanf' declared maximum
field width.  Because 'sscanf' string parsing is rarely used in this code, only two
occurrences of this issue appear in mainstream code and only a few more occur in special
purpose standalone utilities only used in controlled environments (two such utilities,
both found in the 'dbupdate' source subdirectory are also repaired here).

The mainstream occurrence that exposed these inconsistencies occurred in code used to
parse OpenVision file and device specification strings (see changes to backend/PFrun.cpp
and kernel/VkSocketAddress.cpp) and were revealed when Vision's 'asFileContents' message
was used to read a file with a path name longer than roughly 64-bytes (the maximum size
of the URL prefix that could be prepended to that path).  As a practical matter, the
structure of the parser code and the size and layout of the intermediate buffers used
to receive parsed data ensured that no actual buffer overruns occurred outside of the
parser's internal buffer space and that the contents of no internally overrun buffer
were ever examined by the parser, despite the technically accurate detection of the
buffer overruns corrected here.

MichaelJCaruso committed:
Change class Vxa::VCallAgent's designation from CONCRETE to ABSTRACT.

MichaelJCaruso committed:
Remove harmless '&' left during buffer overrun fixes of commit 3c249b4.

MichaelJCaruso committed:
Merge branch 'master' into solaris-11.3-x86

y this merge is necessary,

MichaelJCaruso committed:
Merge branch 'solaris-11.3-x86' into macport-plus

MichaelJCaruso committed:
Merge branch 'master' into macport

MichaelJCaruso committed:
Merge branch 'macport' into macport-plus

MichaelJCaruso committed:
Merge branch 'master' of github.com:MichaelJCaruso/vision

MichaelJCaruso committed:
Merge branch 'master' (solaris-11.3) into macport

MichaelJCaruso committed:
Merge branch 'master' ('macport') into macport-plus

MichaelJCaruso committed:
Unify Solaris Sparc and Intel make recipes.

The make rules developed to support Solaris x86 builds also work for
Solaris Sparc builds.  Use them for both platforms, deleting the now
subsumed M_SunOS_i386 (aka x86) definitions.

MichaelJCaruso committed:
Merge branch 'solaris-11.3-x86' into macport

Conflicts:
	software/src/8.0/src/M_SunOS_i386/make.CC.lib

MichaelJCaruso committed:
Merge branch 'macport' into macport-plus
